### PR TITLE
Fixes #7347: Add rudder-dev revert

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://www.rudder-project.org/redmine/issues/7347

### CODE REVIEW

**Comment by VinceMacBuche** [comment]
PR updated

File: scripts/rudder-dev/rudder-dev [hunk 1]
```diff
@@ -789,6 +790,17 @@ def get_pr_source(pr_url):
   pr = github_request(url, None, pr_url)
   return (pr['head']['repo']['git_url'], pr['head']['ref'])
 
+# get PR merge commit
+def get_pr_merge_commit(pr_url):
+  url = "https://api.github.com/repos/Normation/{repo}/issues/{pr_id}/events"
+  pr_events = github_request(url, None, pr_url)
+  pr_merged = [ pr for pr in pr_events if pr["event"] == "merged" ]
+  if len(pr_merged) == 0:
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: shouldn't it be !=1 ?

File: scripts/rudder-dev/rudder-dev [hunk 2]
```diff
@@ -1269,6 +1281,22 @@ def merge_all(strategy=None, automatic=False):
   for branch in get_versions(lifecycle)[:-1]:
     merge_to_next(branch, strategy, automatic)
 
+# Revert commit from ticket passed as parameter, use retarget to keep changes on next branch  
+def revert(ticket, retarget = False):
+  (ticket_id, internal) = parse_ticket_id(ticket)
+  info = get_ticket_info(ticket_id, internal, False)
+  if retarget:
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: Add a comment to say why we need it and why only for --retarget

File: scripts/rudder-dev/rudder-dev [hunk 3]
```diff
@@ -1269,6 +1281,22 @@ def merge_all(strategy=None, automatic=False):
...
+# Revert commit from ticket passed as parameter, use retarget to keep changes on next branch  
+def revert(ticket, retarget = False):
+  (ticket_id, internal) = parse_ticket_id(ticket)
+  info = get_ticket_info(ticket_id, internal, False)
+  if retarget: 
+    merge_to_next(info['version'])
+  pull(info['version'])
+  if 'pr' in info and info['pr'] != '':
+    commit = get_pr_merge_commit(info['pr'])
+  else:
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: An error message would be welcome

File: scripts/rudder-dev/rudder-dev [hunk 4]
```diff
@@ -17,6 +17,7 @@ Usage:
     rudder-dev [-f|--force] rebase [--base=<ticket_id>] [<PR_comment>]
     rudder-dev [-f|--force] retarget [<target_version>]
     rudder-dev [-f|--force] takeover <ticket_id>
+    rudder-dev [-f|--force] revert <ticket_id> [--retarget]
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: Add a new section below for the new command

File: scripts/rudder-dev/rudder-dev [hunk 5]
```diff
@@ -789,6 +790,17 @@ def get_pr_source(pr_url):
...
 
+# get PR merge commit
+def get_pr_merge_commit(pr_url):
+  url = "https://api.github.com/repos/Normation/{repo}/issues/{pr_id}/events"
+  pr_events = github_request(url, None, pr_url)
+  pr_merged = [ pr for pr in pr_events if pr["event"] == "merged" ]
+  if len(pr_merged) == 0: 
+    print("Can't find merge commit for pull request")
+    exit(12)
+  pprint(pr_merged)
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: pprint ?

**Comment by VinceMacBuche** [comment]
PR updated

File: scripts/rudder-dev/rudder-dev (hunk 1 above)
Comment (VinceMacBuche) [inline scripts/rudder-dev/rudder-dev]: Not sure if it's possible to have more that one 'merged' event ... I guess there should be only one and should be as you said!

**Comment by VinceMacBuche** [comment]
PR updated

File: scripts/rudder-dev/rudder-dev [hunk 6]
```diff
@@ -114,6 +115,15 @@ TAKEOVER
...
     ex: rudder-dev takeover 1234
 
+REVERT
+    - If the ticket has no pull-request, abandon
+    - If retarget, ensure we have everything merged
+    - Find merge commit of pull request
+    - Revert pull request merge commit
+    - if retarget, merge it to next branch with ours strategy (keep changes in next branch)
+    ex: rudder-dev revert 1234
+
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: There are 2 blank lines instead of one

File: scripts/rudder-dev/rudder-dev [hunk 7]
```diff
@@ -789,6 +799,16 @@ def get_pr_source(pr_url):
   pr = github_request(url, None, pr_url)
   return (pr['head']['repo']['git_url'], pr['head']['ref'])
 
+# get PR merge commit
```
Comment (peckpeck) [inline scripts/rudder-dev/rudder-dev]: very useful comment

**Pull request merged**